Cheatsheet: Segregated codes based on type

ojasaklechayt opened the pull request:
With the reference to the issue: #12934 and it's task number 3 i.e.

- [x] Members of address, type, etc. should be in a separate section, not under "Global Variables".

I have segregated the code snippets with reference to the Solidity docs.

ojasaklechayt committed:
Cheatsheet segregated based on type

ojasaklechayt commented:
@cameel

ojasaklechayt commented:
@leonardoalt can you please check?

cameel commented on docs/cheatsheet.rst:
```diff
@@ -6,12 +6,13 @@ Cheatsheet
+
 .. include:: types/operator-precedence-table.rst
 
 .. index:: assert, block, coinbase, difficulty, prevrandao, number, block;number, timestamp, block;timestamp, msg, data, gas, sender, value, gas price, origin, revert, require, keccak256, ripemd160, sha256, ecrecover, addmod, mulmod, cryptography, this, super, selfdestruct, balance, codehash, send
```
These index entries should be split too so that each one is placed next to section that talks about its topic.

ojasaklechayt replied:
Not sure about this. It would be great it you can brief me more about it.

cameel replied:
The [`index` directive](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#index-generating-markup) adds entries to [keyword index](https://docs.soliditylang.org/en/latest/genindex.html). As it is now, all of them are pointing at a section called `ABI Encoding and Decoding Functions`, which is wrong. You need to have separate `index` directive for each section, with entries matching what's actually inside that section.

ojasaklechayt replied:
Added the indices.

cameel commented on docs/cheatsheet.rst:
```diff
@@ -25,10 +26,23 @@ Global Variables
 - ``abi.encodeWithSignature(string memory signature, ...) returns (bytes memory)``: Equivalent
   to ``abi.encodeWithSelector(bytes4(keccak256(bytes(signature))), ...)``
+  
+Members of bytes
+================
```
Not sure if this will work (please check the preview) but if it does, it would be better this way:
```suggestion
Members of ``bytes``
====================
```

Same for ``string`` below.

ojasaklechayt replied:
Hii!! It's working

ojasaklechayt replied:
I have done the changes 👍

cameel commented on docs/cheatsheet.rst:
```diff
@@ -25,10 +26,23 @@ Global Variables
 - ``abi.encodeWithSignature(string memory signature, ...) returns (bytes memory)``: Equivalent
   to ``abi.encodeWithSelector(bytes4(keccak256(bytes(signature))), ...)``
+  
+Members of bytes
```
I'd group things for `string` and `bytes` in a single section.

ojasaklechayt replied:
Not sure, What should I name it so I just named it " Members of bytes and string "

cameel replied:
Yeah. "Members of `bytes` and `string`" sounds just fine.

cameel commented on docs/cheatsheet.rst:
```diff
@@ -61,15 +82,27 @@ Global Variables
 - ``super``: the contract one level higher in the inheritance hierarchy
 - ``selfdestruct(address payable recipient)``: destroy the current contract, sending its funds to the given address
+
+Members of Address Types
+========================
```
I'd move the section about `string` and `bytes` next to this one so that we have all type members in a single spot.

ojasaklechayt replied:
Done

cameel commented on docs/cheatsheet.rst:
```diff
@@ -61,15 +82,27 @@ Global Variables
 - ``mulmod(uint x, uint y, uint k) returns (uint)``: compute ``(x * y) % k`` where the multiplication is performed
   with arbitrary precision and does not wrap around at ``2**256``. Assert that ``k != 0`` starting from version 0.5.0.
+  
+Contract Related
```
```suggestion
Contract-related
```

ojasaklechayt replied:
Done

cameel commented on docs/cheatsheet.rst:
```diff
@@ -44,14 +58,21 @@ Global Variables
 - ``tx.gasprice`` (``uint``): gas price of the transaction
 - ``tx.origin`` (``address``): sender of the transaction (full call chain)
+
+Error Handling
```
```suggestion
Error Reporting
```

ojasaklechayt replied:
Done

cameel commented on docs/cheatsheet.rst:
```diff
@@ -61,15 +89,17 @@ Global Variables
+================
+
 - ``this`` (current contract's type): the current contract, explicitly convertible to ``address`` or ``address payable``
 - ``super``: the contract one level higher in the inheritance hierarchy
```
One more thing, I noticed that we don't have `super` listed on the "Units and Globally Available Variables" page. Can you add it there?

ojasaklechayt replied:
Done

cameel commented on docs/units-and-global-variables.rst:
```diff
@@ -329,6 +329,9 @@ Contract Related
 
 Furthermore, all functions of the current contract are callable directly including the current function.
 
+``super``
+    The contract one level higher in the inheritance hierarchy
```
Can you put it next to `this`?

ojasaklechayt replied:
Done

cameel commented on docs/cheatsheet.rst:
```diff
@@ -61,15 +97,19 @@ Global Variables
-  returns ``false`` on failure
-- ``<address payable>.transfer(uint256 amount)``: send given amount of Wei to :ref:`address`, throws on failure
+
+Type Information
+================
```
I'd add index entries for all the properties here: `type; name`, `type; creationCode`, `type; runtimeCode`, and so on.

ojasaklechayt replied:
I have added it

cameel commented on docs/cheatsheet.rst:
```diff
@@ -25,10 +26,34 @@ Global Variables
+  returns ``false`` on failure
+- ``<address payable>.transfer(uint256 amount)``: send given amount of Wei to :ref:`address`, throws on failure
+
+.. index:: block, coinbase, difficulty, prevrandao, number, block;number, timestamp, block;timestamp, msg, data, gas, sender, value, gas price, origin
+
+Block and Transaction Properties
+================================
```
Here also some of them are missing from the index. Can you add them?

For ones that are properties like `block.basefee` or `tx.origin` we can use the `;` notation: `block; basefee`, `tx; origin`. This means that the index will contain a single entry for `block`/`tx` and `basefee`/`origin` will be subentries under it (see how it looks like in the index).

ojasaklechayt replied:
Most of them are already there so I have added remaining once

cameel commented on docs/cheatsheet.rst:
```diff
@@ -25,10 +26,34 @@ Global Variables
 - ``string.concat(...) returns (string memory)``: :ref:`Concatenates variable number of
   arguments to one string array<string-concat>`
+
+.. index:: balance, codehash, send
+
+Members of Address Types
+========================
```
Here we're missing `transfer` and `code` in the index. Also we could use the `address; transfer` notation to make them subentries.

ojasaklechayt replied:
Done

cameel commented on docs/cheatsheet.rst:
```diff
@@ -25,10 +26,34 @@ Global Variables
 - ``abi.encodeWithSignature(string memory signature, ...) returns (bytes memory)``: Equivalent
   to ``abi.encodeWithSelector(bytes4(keccak256(bytes(signature))), ...)``
+
+Members of ``bytes`` and  ``string``
+====================================
```
I'd add index entries for `bytes; concat` and `string; concat` here.

ojasaklechayt replied:
done

cameel commented on docs/cheatsheet.rst:
```diff
@@ -6,12 +6,13 @@ Cheatsheet
 .. include:: types/operator-precedence-table.rst
 
-.. index:: assert, block, coinbase, difficulty, prevrandao, number, block;number, timestamp, block;timestamp, msg, data, gas, sender, value, gas price, origin, revert, require, keccak256, ripemd160, sha256, ecrecover, addmod, mulmod, cryptography, this, super, selfdestruct, balance, codehash, send
+.. index:: cryptography
```
Weird that we have "cryptography" as an index entry :) But in any case, it would fit more on the "Mathematical and Cryptographic Functions" section.

ojasaklechayt replied:
Done

cameel reviewed (commented):
Some index tweaks are still needed. Other than that looks good!

By the way, before we merge this, we'll want the commits to be squashed into something more sensible (all of this should probably be just one or two commits). And we never merge `develop` into PRs. We rebase PRs on top of it instead.

ojasaklechayt commented:
> Some index tweaks are still needed. Other than that looks good!
> 
> By the way, before we merge this, we'll want the commits to be squashed into something more sensible (all of this should probably be just one or two commits). And we never merge `develop` into PRs. We rebase PRs on top of it instead.

Okay so I will create another PR with fewer commit and will tag you there

cameel commented:
> My pr is using patch-3 so it might not be the problem merging I guess

Wait, why? It's perfectly fine to just rebase this one.

ojasaklechayt commented:
> > My pr is using patch-3 so it might not be the problem merging I guess
> 
> Wait, why? It's perfectly fine to just rebase this one.
> 
> We rebase PRs all the time. Creating a new one each time you do that would be pretty spammy :)

Oh Okay, so I searched about rebasing PR, Is this like changing my pr from patch-3 to develop ? Rebasing is kind of new to me 😅

cameel commented:
> Oh yeah okay I got it. So, instead of directly merging into develop branch which is the main branch, we shall upload it into another branch which may be gradually be merged into the main branch. So, which branch shall I merge into ?

No, it's exactly the opposite. Gradual merging is what you did here (just in the other direction - from `develop` into your branch) and we do not normally do that. Instead we rebase the branch on top of `develop` and merge it into it in one go.

This way there are invisible changes in history coming from conflict resolutions inside merge commits. Every conflict resolution is done explicitly as a part of one of your commits.

ojasaklechayt commented:
@matheusaaguiar

ojasaklechayt commented:
@cameel any updates regarding this pr ?

matheusaaguiar commented on docs/cheatsheet.rst:
```diff
@@ -25,10 +26,36 @@ Global Variables
 - ``string.concat(...) returns (string memory)``: :ref:`Concatenates variable number of
   arguments to one string array<string-concat>`
+
+.. index:: balance, codehash, send, address; code, address; transfer
```
```suggestion
.. index:: address; balance, address; codehash, address; send, address; code, address; transfer
```
Why not?

matheusaaguiar commented on docs/cheatsheet.rst:
```diff
@@ -25,10 +26,36 @@ Global Variables
+
+.. index:: balance, codehash, send, address; code, address; transfer
+
+Members of Address Types
```
```suggestion
Members of Address Type
```

matheusaaguiar commented on docs/cheatsheet.rst:
```diff
@@ -25,10 +26,36 @@ Global Variables
+  returns ``false`` on failure
+- ``<address payable>.transfer(uint256 amount)``: send given amount of Wei to :ref:`address`, throws on failure
+
+.. index:: block, block; basefree,block; chainid,  coinbase, difficulty, prevrandao, number, block;number, timestamp, block;timestamp, msg, data, gas, sender, value, gas price, origin
```
```suggestion
.. index:: block, block; basefree, block; chainid, coinbase, difficulty, prevrandao, number, block; number, timestamp, block; timestamp, msg, data, gas, sender, value, gas price, origin
```

matheusaaguiar commented on docs/cheatsheet.rst:
```diff
@@ -44,14 +71,25 @@ Global Variables
+
+.. index:: assert, require, revert
+
+Error Reporting
```
```suggestion
Error Handling
```
I think something like that is more fitting, although error reporting is also part of the effects of these. But to me seems like `assert`, for example, is more about logical assumptions and exceptions (absurd values/states) than reporting per se.

matheusaaguiar replied:
Just saw that @cameel suggested the opposite before. So feel free to ignore this.

cameel replied:
Yeah, `assert` is not for reporting errors, but also none of these functions is actually handling errors. I guess the current section name does give a wrong impression though. Changed to `Validations and Assertions`.

matheusaaguiar commented on docs/cheatsheet.rst:
```diff
@@ -61,15 +99,21 @@ Global Variables
+
+.. index:: this, super, selfdestruct
+
+Contract-related
```
```suggestion
Contract Related
```

matheusaaguiar replied:
Feel free to ignore this as well.

matheusaaguiar commented on docs/units-and-global-variables.rst:
```diff
@@ -311,14 +311,17 @@ For more information, see the section on :ref:`address`.
+    The current contract, explicitly convertible to :ref:`address`
 
+``super``
+    The contract one level higher in the inheritance hierarchy
```
```suggestion
    The parent contract which the current contract inherits from
```

cameel replied:
I'm not sure if that's really better. Both seem to imply that there's just one contract while `super` is a really weird thing that can refer to inherited stuff from multiple contracts from multiple levels.

cameel replied:
I changed it to "a contract", which is better, but we should find better wording at some point.

matheusaaguiar commented on docs/units-and-global-variables.rst:
```diff
@@ -311,14 +311,17 @@ For more information, see the section on :ref:`address`.
-.. index:: this, selfdestruct
+.. index:: this, selfdestruct, super
 
 Contract Related
```
@cameel, Should this be changed to `Contract-related` to be consistent with the section name in the cheatsheet?

cameel replied:
Sure, why not. That would be the correct way to phrase it. It looks weird without a dash.

matheusaaguiar commented:
@ojasaklechayt the commits could be cleaned a bit, especially the `Merge Branch` ones.
I would like to kindly recommend this as a resource for better understanding of `git`: https://git-scm.com/book/en/v2
It was very helpful to me when I started working with the `tool` and made everything much more clear.

ojasaklechayt commented:
@cameel @matheusaaguiar is it better now? I tried squashing previous commits but it somehow didn't worked? But then I got how to do rebase. Is it correct. I have been trying this for long time but somehow it isn't happening as I mostly use github website to upload 😅

ojasaklechayt commented:
> @cameel @matheusaaguiar is it better now? I tried squashing previous commits but it somehow didn't worked? But then I got how to do rebase. Is it correct. I have been trying this for long time but somehow it isn't happening as I mostly use github website to upload 😅

@cameel @matheusaaguiar

cameel commented on docs/cheatsheet.rst:
```diff
@@ -6,12 +6,13 @@ Cheatsheet
 .. include:: types/operator-precedence-table.rst
 
-.. index:: assert, block, coinbase, difficulty, prevrandao, number, block;number, timestamp, block;timestamp, msg, data, gas, sender, value, gas price, origin, revert, require, keccak256, ripemd160, sha256, ecrecover, addmod, mulmod, cryptography, this, super, selfdestruct, balance, codehash, send
+.. include:: abi; decode, abi; encode, abi; encodePacked, abi; encodeWithSelector, abi; encodeCall, abi; encodeWithSignature
```
`include` is for including files. This was supposed to be `index`. Please note that the `b_docs` job is failing because of this. The PR must pass all the CI checks before it can be merged.

cameel commented on docs/units-and-global-variables.rst:
```diff
@@ -311,14 +311,17 @@ For more information, see the section on :ref:`address`.
 
+``super``
+    The parent contract which the current contract inherits from
+    
```
You have some trailing whitespace here. Note that `chk_coding_style` job is failing because of this.

cameel commented on docs/cheatsheet.rst:
```diff
@@ -25,10 +26,36 @@ Global Variables
+  returns ``false`` on failure
+- ``<address payable>.transfer(uint256 amount)``: send given amount of Wei to :ref:`address`, throws on failure
+
+.. index:: block, block; basefree, block; chainid, coinbase, difficulty, prevrandao, number, block; number, timestamp, block; timestamp, msg, data, gas, sender, value, gas price, origin
```
`blockhash` is missing from the index.

cameel replied:
Also, this is generally out of sync with the section below. Some items are included twice (`timestamp`, `number`), some are missing (`gasleft`).

cameel commented:
> is it better now? I tried squashing previous commits but it somehow didn't worked? But then I got how to do rebase. Is it correct. I have been trying this for long time but somehow it isn't happening as I mostly use github website to upload sweat_smile

Unfortunately it isn't. After a rebase you should get a simple, linear, easy to understand history. Whatever you ended up with... was not it :) In general, if you see any commits named `Merge branch ... into ...`, it's not properly rebased.

Anyway, I rebased the branch for you and fixed the small issues I just pointed out. I'm going to merge it now. Thanks for contribution!

ojasaklechayt commented:
> > is it better now? I tried squashing previous commits but it somehow didn't worked? But then I got how to do rebase. Is it correct. I have been trying this for long time but somehow it isn't happening as I mostly use github website to upload sweat_smile
> 
> Unfortunately it isn't. After a rebase you should get a simple, linear, easy to understand history. Whatever you ended up with... was not it :) In general, if you see any commits named `Merge branch ... into ...`, it's not properly rebased.
> 
> Anyway, I rebased the branch for you and fixed the small issues I just pointed out. I'm going to merge it now. Thanks for contribution!

Thankyou so much for the support @cameel. I learned a lot with this pr.

ojasaklechayt commented:
@matheusaaguiar I think your review is remaining
